Coherent sheaves

HechtiDerLachs opened the pull request:
This is a draft on coherent sheaves on `AbsCoveredScheme`s. As an example, we implemented twisting sheaves on projective space and the cotangent sheaf of an arbitrary covered scheme. See the test file for samples.

HechtiDerLachs committed:
Make SubModuleOfFreeModule attributable.

HechtiDerLachs committed:
Make SubModuleOfFreeModule attributable.

HechtiDerLachs committed:
Make SubModuleOfFreeModule attributable.

HechtiDerLachs committed:
Implement the SubModuleOfFreeModule-level for MPolyQuos.

HechtiDerLachs committed:
Implement the SubModuleOfFreeModule-level for MPolyQuos.

HechtiDerLachs committed:
Implement the SubModuleOfFreeModule-level for MPolyQuos.

HechtiDerLachs committed:
Include tests.

HechtiDerLachs committed:
Include tests.

HechtiDerLachs committed:
Include tests.

HechtiDerLachs committed:
Temporary fix of division in quotient rings.

HechtiDerLachs committed:
Temporary fix of division in quotient rings.

HechtiDerLachs committed:
Merge branch 'mpolyquo_modules' into coherent_sheaves

HechtiDerLachs committed:
Fix the matrix function to accept morphisms with base change and repare equality check of submodules.

HechtiDerLachs committed:
Fix the matrix function to accept morphisms with base change and repare equality check of submodules.

HechtiDerLachs committed:
Fix the matrix function to accept morphisms with base change and repare equality check of submodules.

HechtiDerLachs committed:
Fix the matrix function to accept morphisms with base change and repare equality check of submodules.

HechtiDerLachs committed:
First draft of coherent sheaves.

HechtiDerLachs committed:
Introduce extra type parameter for base change in module homomorphisms and fix composition.

HechtiDerLachs committed:
Introduce extra type parameter for base change in module homomorphisms and fix composition.

HechtiDerLachs committed:
Introduce extra type parameter for base change in module homomorphisms and fix composition.

HechtiDerLachs committed:
Introduce extra type parameter for base change in module homomorphisms and fix composition.

HechtiDerLachs committed:
Add tests from the issue #1806.

HechtiDerLachs committed:
Add tests from the issue #1806.

HechtiDerLachs committed:
Add tests from the issue #1806.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Add indentation in the tests.

HechtiDerLachs committed:
Add indentation in the tests.

HechtiDerLachs committed:
Add indentation in the tests.

HechtiDerLachs committed:
Explain type parameter for morphisms of modules.

HechtiDerLachs committed:
Explain type parameter for morphisms of modules.

HechtiDerLachs committed:
Explain type parameter for morphisms of modules.

HechtiDerLachs committed:
Merge branch 'mpolyquo_modules' into coherent_sheaves

HechtiDerLachs committed:
Fix maps to take into account a base ring map.

HechtiDerLachs committed:
Fix maps to take into account a base ring map.

HechtiDerLachs committed:
Fix maps to take into account a base ring map.

HechtiDerLachs committed:
Make base change of free modules remember the symbols.

HechtiDerLachs committed:
Make base change of free modules remember the symbols.

HechtiDerLachs committed:
Make base change of free modules remember the symbols.

HechtiDerLachs committed:
Implement twisting sheaves and the cotangent sheaf.

HechtiDerLachs committed:
Fix constructor of cotangent sheaf.

HechtiDerLachs committed:
Repair dispatch for mapping of module elements.

HechtiDerLachs committed:
Repair dispatch for mapping of module elements.

HechtiDerLachs committed:
Repair dispatch for mapping of module elements.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Add generic method to create a submodule by multiplication with an ideal.

HechtiDerLachs committed:
Add generic method to create a submodule by multiplication with an ideal.

HechtiDerLachs committed:
Add generic method to create a submodule by multiplication with an ideal.

HechtiDerLachs committed:
Minor fixes in coordinates for localizations.

HechtiDerLachs committed:
Minor fixes in coordinates for localizations.

HechtiDerLachs committed:
Minor fixes in coordinates for localizations.

HechtiDerLachs committed:
Add abstract type for coherent sheaves and special type for hom-sheaves.

HechtiDerLachs committed:
Add tests.

HechtiDerLachs committed:
Minor fixes and constructor for dual.

HechtiDerLachs commented on experimental/Schemes/CoherentSheaves.jl:
```diff
@@ -0,0 +1,461 @@
+        images_f = [sum([B[i][j]*cod_res(phi_map(f[j])) for j in 1:length(f)]) for i in 1:length(B)]
+        psi = hom(F(U), G(U), images_f)
+        push!(images, homomorphism_to_element(MU, psi))
+      end
```
@jankoboehm : I implemented the restriction of homomorphisms to open subsets here. But I'm not sure, I am using your machinery as intended. In particular the conversion of elements to homomorphisms and back seems to be a bit tedious. Do you provide something like "induced hom from left" and "induced hom from right" for "composition" of Hom-modules with homomorphisms to the domain and from the codomain?

jankoboehm replied:
The construction from what we have is not so difficult and too many shortcuts might in the end be difficult to remember and confusing, but I think it is a good idea to introduce it to provide all the automatisms of Hom modules. We have lift_homomorphism_contravariant and lift_homomorphism_contravariant, though not exactly what you want.

HechtiDerLachs replied:
Ok, I'll have a look. But then you say that my solution is in the scope of how you intended users to handle the modules? 

If not, feel free to propose another solution here!

HechtiDerLachs committed:
Add Cartier divisors.

HechtiDerLachs committed:
Add tests.

HechtiDerLachs committed:
Add line bundles constructed from Cartier divisors.

HechtiDerLachs committed:
Minor fix.

HechtiDerLachs committed:
Change signature of internal production and restriction functions for sheaves.

HechtiDerLachs committed:
Adjust the coherent sheaves.

HechtiDerLachs committed:
Clean up preimage functionality.

HechtiDerLachs committed:
Clean up preimage functionality.

HechtiDerLachs committed:
Clean up preimage functionality.

HechtiDerLachs committed:
Implement pushforwards of sheaves along closed embeddings.

HechtiDerLachs committed:
Clean up the ideal sheaves for the new internal signatures.

HechtiDerLachs committed:
Almost fix dispatch of methods.

HechtiDerLachs committed:
Almost fix dispatch of methods.

HechtiDerLachs committed:
Almost fix dispatch of methods.

HechtiDerLachs committed:
Implement additional required coordinate method.

HechtiDerLachs committed:
Implement additional required coordinate method.

HechtiDerLachs committed:
Implement additional required coordinate method.

HechtiDerLachs committed:
Repair method.

HechtiDerLachs committed:
Fix pushforward sheaves.

HechtiDerLachs committed:
Add tests.

HechtiDerLachs committed:
Merge branch 'master' into mpolyquo_modules

HechtiDerLachs committed:
Merge branch 'master' into coherent_sheaves

HechtiDerLachs committed:
Do the changes again which were reverted by the merge.

HechtiDerLachs committed:
Do the changes again which were reverted by the merge.

HechtiDerLachs committed:
Do the changes again which were reverted by the merge.

HechtiDerLachs committed:
Repair the faulty kernel computation over MPolyQuoLocalizedRings.

HechtiDerLachs committed:
Repair the faulty kernel computation over MPolyQuoLocalizedRings.

HechtiDerLachs committed:
Add tests for the fixed issue.

HechtiDerLachs committed:
Add tests for the fixed issue.

HechtiDerLachs committed:
Merge branch 'mpolyquo_modules' into coherent_sheaves

HechtiDerLachs committed:
Remove superfluous type parameters as they were cluttering the screen completely.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Implement printing.

HechtiDerLachs committed:
Minor fix.

HechtiDerLachs committed:
Fix restriction to principal open subsets.

HechtiDerLachs committed:
Implement the kernel for FreeMod -> SubQuo manually.

HechtiDerLachs committed:
Implement the kernel for FreeMod -> SubQuo manually.

HechtiDerLachs committed:
Implement the kernel for FreeMod -> SubQuo manually.

HechtiDerLachs committed:
Implement free resolutions.

HechtiDerLachs committed:
Implement free resolutions.

HechtiDerLachs committed:
Implement free resolutions.

HechtiDerLachs committed:
Add a truely generic method for free_resolution and restrict the existing one to polynomial rings.

HechtiDerLachs committed:
Add a truely generic method for free_resolution and restrict the existing one to polynomial rings.

HechtiDerLachs committed:
Add a truely generic method for free_resolution and restrict the existing one to polynomial rings.

HechtiDerLachs committed:
Introduce types for sections of presheaves.

HechtiDerLachs committed:
Implement sections in coherent sheaves and global sections of twisting sheaves.

HechtiDerLachs committed:
Carry over the fineprint of previously existing method free_resolution_via_kernel.

HechtiDerLachs committed:
Carry over the fineprint of previously existing method free_resolution_via_kernel.

HechtiDerLachs committed:
Merge branch 'mpolyquo_modules' into coherent_sheaves

HechtiDerLachs committed:
Make iszero generic for module elements and add ambient_representative for free module elements.

HechtiDerLachs committed:
WIP on extensions of sections.

HechtiDerLachs committed:
Fix doctests.

HechtiDerLachs committed:
Going back to the old version of projection maps to quotient rings.

HechtiDerLachs committed:
Fix doctests again.

HechtiDerLachs committed:
Fix doctests again after rebase on master.

HechtiDerLachs committed:
Merge branch 'mpolyquo_modules' into coherent_sheaves

HechtiDerLachs committed:
WIP on PullbackSheaves.

HechtiDerLachs committed:
Halfway through implementing PullbackSheaf.

HechtiDerLachs committed:
Towards caching of restrictions inside the codomain.

This is important for the garbage collector: In the long
run, a sheaf must be able to forget its values on deprecated
open subsets. If the restriction maps were stored directly
in the sheaf, this would block forgetting because of the
references.

HechtiDerLachs committed:
Some fixes for boundary cases with empty lists.

HechtiDerLachs committed:
Make tests run again for ideal sheaves.

HechtiDerLachs committed:
Rebase coherent sheaves on the new caching framework.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Merge branch 'master' into coherent_sheaves

HechtiDerLachs committed:
Repair the ideal sheaves.

HechtiDerLachs committed:
Repair iszero test for SubQuoElems.

simonbrandhorst commented:
Why does this pr change doctests all over the place? This looks fishy.

HechtiDerLachs commented:
Yes, it's fishy. That is because I had to run the doctests with `:fix` once after changing the type of the map for the projections to quotient rings and that affected a lot of files. These changes have been reverted, but the doctests are still in the old state. Now, when I run them with `:fix` again, I will still mess things up, because I have the wrong julia version installed. I will take care of this, eventually.

HechtiDerLachs commented:
The other question is: How can I clean up this huge mess in the commit-history? 

The background is that I had to develop two PRs in parallel, this one and the one on modules over `MPolyQuo`s. To this end, I shifted a lot of commits back and forth between the two using `git cherry-pick`. But that makes them appear as different commits in history despite doing the exact same changes. A rebase on master seemed impossible here, so I did a merge. But that leaves us with the duplication mess above. Any suggestions how to avoid this in the future is most welcome (via the usual channels). 

My apologies for what you see above. I guess, we have to squash this one eventually.

simonbrandhorst commented:
This PR is just open to discuss on coherent sheaves. It is not planned to merge this.
